org.eclipse.xtext.xtext.ui.tests/src/org/eclipse/xtext/xtext/ui/editor/quickfix/XtextGrammarQuickfixTest.xtend

```diff
@@ -24,27 +24,156 @@ import static org.eclipse.xtext.xtext.XtextConfigurableIssueCodes.SPACES_IN_KEYW
 		'''
 			grammar org.xtext.example.mydsl.MyDsl with org.eclipse.xtext.common.Terminals
 			
 			generate myDsl "http://www.xtext.org/example/mydsl/MyDsl"
 			
 			Model:
-				greetings+=Greeting*;
+				a = ABC;
+			
+			terminal fragment ABC:
+				'a';
+		'''.testQuickfixesOn(
+			"org.eclipse.xtext.grammar.InvalidTerminalFragmentRuleReference",
+			new Quickfix("Convert terminal fragment to terminal rule", "Convert terminal fragment to terminal rule", '''
+				grammar org.xtext.example.mydsl.MyDsl with org.eclipse.xtext.common.Terminals
+				
+				generate myDsl "http://www.xtext.org/example/mydsl/MyDsl"
+				
+				Model:
+					a = ABC;
+				
+				terminal ABC:
+					'a';
+			''')
+		);
+	}
+
+	@Test def test_fix_invalid_hidden_token() {
+		'''
+			grammar org.xtext.example.mydsl.MyDsl with org.eclipse.xtext.common.Terminals
+			
+			generate myDsl "http://www.xtext.org/example/mydsl/MyDsl"
+			
+			Model hidden(AnotherModel):
+				a = ID;
+			
+			AnotherModel:
+				b = ID;
+		'''.testQuickfixesOn(
+			"org.eclipse.xtext.grammar.InvalidHiddenToken",
+			new Quickfix("Remove hidden token definition", "Remove hidden token definition", '''
+				grammar org.xtext.example.mydsl.MyDsl with org.eclipse.xtext.common.Terminals
+				
+				generate myDsl "http://www.xtext.org/example/mydsl/MyDsl"
+				
+				Model hidden():
+					a = ID;
+				
+				AnotherModel:
+					b = ID;
+			''')
+		);
+	}
+
+	@Test def test_fix_missing_rule() {
 		'''
-		.testQuickfixesOn("org.eclipse.xtext.grammar.UnresolvedRule", new Quickfix("Create rule 'Greeting'", "Create rule 'Greeting'", '''
 			grammar org.xtext.example.mydsl.MyDsl with org.eclipse.xtext.common.Terminals
 			
 			generate myDsl "http://www.xtext.org/example/mydsl/MyDsl"
 			
 			Model:
 				greetings+=Greeting*;
-			
-			Greeting:
+		'''.testQuickfixesOn(
+			"org.eclipse.xtext.grammar.UnresolvedRule",
+			new Quickfix("Create rule 'Greeting'", "Create rule 'Greeting'", '''
+				grammar org.xtext.example.mydsl.MyDsl with org.eclipse.xtext.common.Terminals
+				
+				generate myDsl "http://www.xtext.org/example/mydsl/MyDsl"
+				
+				Model:
+					greetings+=Greeting*;
+				
+				Greeting:
+					
+				;
+			'''),
+			new Quickfix("Create enum rule 'Greeting'", "Create enum rule 'Greeting'", '''
+				grammar org.xtext.example.mydsl.MyDsl with org.eclipse.xtext.common.Terminals
+				
+				generate myDsl "http://www.xtext.org/example/mydsl/MyDsl"
+				
+				Model:
+					greetings+=Greeting*;
+				
+				enum Greeting:
+					
+				;
+			'''),
+			new Quickfix("Create terminal 'Greeting'", "Create terminal 'Greeting'", '''
+				grammar org.xtext.example.mydsl.MyDsl with org.eclipse.xtext.common.Terminals
+				
+				generate myDsl "http://www.xtext.org/example/mydsl/MyDsl"
+				
+				Model:
+					greetings+=Greeting*;
+				
+				terminal Greeting:
+					
+				;
+			'''),
+			new Quickfix("Create terminal fragment 'Greeting'", "Create terminal fragment 'Greeting'", '''
```

Comment:
@szarnekow Sorry, I do not understand "but I prefer to adjust this in a subsequent step to move this change forward". Could you please clarify.

Reply:
I'd like to merge this change as-is but create a follow-up task to avoid the quickfix to create a terminal fragment where the fragment usage would be illegal.

Reply:
@szarnekow Tx for clarifying. @cdietrich Would you mind merging the PR if you are ok with it. Tx.